runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -0,0 +1,98 @@
+package org.corfudb.runtime;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.Message;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.CorfuStoreEntry;
+import org.corfudb.runtime.collections.IsolationLevel;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.RetryExhaustedException;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.runtime.LogReplication.ReplicationStatusKey;
+import org.corfudb.runtime.LogReplication.ReplicationStatusVal;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.RetryNeededException;
+import java.lang.reflect.InvocationTargetException;
+import java.time.Duration;
+import java.util.List;
+
+import static org.corfudb.runtime.view.ObjectsView.REPLICATION_STATUS_TABLE;
+import static org.corfudb.runtime.view.TableRegistry.CORFU_SYSTEM_NAMESPACE;
+
+/**
+ * LogReplication code resides in the infrastructure package.  Adding a dependency from this package(runtime) to
+ * infrastructure introduces a circular dependency.  This class defines LR-specific utility methods required in
+ * runtime.  Note that these methods are unique and not duplicated from infrastructure.
+ */
+@Slf4j
+public class LogReplicationUtils {
+
+    /**
+     * Invokes a callback to get the intended subscription timestamp from the client subscribing to the listener.  If
+     * this timestamp is within an ongoing snapshot sync, requests a new timestamp.  If the timestamp is outside a
+     * snapshot sync, invokes a callback on the client to initialize its view of the LR(and any other client) tables.
+     * @param streamListener listener to which an external client has subscribed
+     * @param corfuStore corfu store
+     * @return
+     */
+    public CorfuStoreMetadata.Timestamp performInitialMergeAndGetSubscriptionTimestamp(
+            LRMultiNamespaceListener streamListener, CorfuStore corfuStore) {
+        int numRetries = 10;
+        try {
+            IRetry.build(ExponentialBackoffRetry.class, () -> {
+                CorfuStoreMetadata.Timestamp timestamp;
+                try {
+                    log.info("Get the subscription timestamp via listener callback...");
+                    timestamp = streamListener.getSubscriptionTimestamp();
+                    // Check that a snapshot sync was not ongoing at subscriptionTimestamp.
+                    // If ongoing snapshot sync, retry
+                    if (checkSnapshotSyncOngoing(timestamp, corfuStore)) {
+                        log.info("Snapshot sync is in progress at subscription timestamp {}.  Attempt to get another " +
```

Comment:
Attempt to get another timestamp? shouldn't we wait for the full sync to complete?.. it will be attempted on the completion of the first full sync if a second is encountered.